File: BT/FindTheLargestNumberInARotatedList/FindTheLargestNumberInARotatedList.java
import java.util.*;

/*
[6,7,8,1,4]

[6,7,8,9,4]

[5,1,2,3,4]
[6,5,1,2,3,4]

[1,2,3,4,5]

*/
class Solution {
    public int solve(int[] arr) {
        if (arr.length == 0) return 0;
        int maxNum = arr[0];
        int l = 0, r = arr.length-1;
        while (l <= r) {
            int m = (l + r) / 2;
            maxNum = Math.max(maxNum, arr[m]);

            // Bottom half is sorted
            if (arr[l] < arr[m]) {
                l = m + 1;
            } else { // Top half is sorted
                // Record the highest
                maxNum = Math.max(maxNum, arr[r]);
                r = m - 1;
            }
        }
        return maxNum;
    }
}

File: BT/FindTheLargestNumberInARotatedList/FindTheLargestNumberInARotatedList.ts
class Solution {
    solve(arr: Array<number>): number {
        if (arr.length === 0) return -1;

        let l = 0,
            r = arr.length-1;
        
        let maxInt = arr[0];
        while (l <= r) {
            const m = Math.floor((l+r)/2);

            if (arr[l] <= arr[m]) {
                maxInt = Math.max(arr[m], maxInt);
                l = m+1;
            } else {
                maxInt = Math.max(arr[r], maxInt);
                r = m-1;
            }
        }
        return maxInt
    }
}
